### ISSUE TITLE
Add config for no-unsafe-optional-chaining to catch unparenthesized unsafe chains

### ISSUE THREAD

**Issue by DillonSadofsky** [eslint/eslint#15789]
### What rule do you want to change?

no-unsafe-optional-chaining

### What change to do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
a?.b.c
a?.b()
```

### What does the rule currently do for this code?

No warnings/errors

### What will the rule do after it's changed?

Warn/error based on rule configuration

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

I welcome some input as to why no-unsafe-optional-chaining treats:
```js
(a?.b).c
```
as an error but
```js
a?.b.c
```
is not, despite the fact that they are both equally capable of throwing an error.  I first started investigated adding `no-unsafe-optional-chaining` because I recently made the mistake of writing some code:
```ts
if ((err as GeneralError)?.is("REGION_NOT_SUPPORTED")) {
```
instead of:
```ts
if ((err as GeneralError)?.is?.("REGION_NOT_SUPPORTED")) {
```

It seems like any time a reference chain of `x.y.[...].z` has an optional chain all 'subsequent' references should be optional as well, or risk undefined errors.  This rule sounded like what I wanted, but it seems to require that I write:
```js
(a?.b)()
```
to get the error, while I will sometimes write it without the parentheses (`a?.b()`) and maybe I'm the only one who would expect this rule to 'trigger', but it seemed somewhat desirable.  I'm curious if there is some background to why it doesn't trigger in that scenario?  The rule page doesn't say.  If this would move too much cheese, I'd welcome a configuration argument that would enable this 'stricter' mode.

Also, somewhat unrelated, but the documentation says that this is part of `eslint:recommended` in the rule page, but I couldn't get it to trigger in my rules that extend that until I explicitly added `'no-unsafe-optional-chaining': [ 'warn' ],` to my `.eslintrc.js`.  I am willing to believe I've done something wrong, but I'm curious if there is some other requirement that I should check in my extension config file?

**Comment by nzakas** [comment]
Thanks for the issue. It’s not clear to me what enhancement you are asking for. Can you be more specific?

For your examples:

```js
(a?.b).c
```

I’m this case. `.c` is always evaluated regardless of the result of `a?.b`, which can be undefined and cause an error, so this is considered unsafe. 

```js
a?.b.c
```

In this case, `.c` is only evaluated if `a?.b` is not undefined. The optional chaining operator short circuits the rest of the expression so this is considered safe. 


Indeed the rule is in the recommended config:
https://github.com/eslint/eslint/blob/413f1d55f0ad05b6fe75bdde6df423253806797d/conf/eslint-recommended.js#L65

The recommended config is different for each version of ESLint, so double-check that you are using the latest version.

**Comment by nzakas** [comment]
> I don't know why it would fire when I explicitly enable it but not if I omit it. 

It’s most likely that another config is disabling it. You can use `—debug` to better see what’s happening.

I’ll need to take a look at these cases in more detail when I can get to a computer. I tried with an online example and I got the short-circuiting behavior. 

The rule itself is very explicit about what is expected to trigger a warning and what is not, and all of the cases you want to trigger a warning are already listed as valid for the rule, so I don’t believe this is a bug but rather an intentional behavior.

**Comment by mdjermanovic** [comment]
> I welcome some input as to why no-unsafe-optional-chaining treats:
> 
> ```js
> (a?.b).c
> ```
> 
> as an error but
> 
> ```js
> a?.b.c
> ```
> 
> is not

Per the [documentation](https://eslint.org/docs/rules/no-unsafe-optional-chaining) for this rule:

> This rule aims to detect some cases where the use of optional chaining doesn't prevent runtime errors. In particular, it flags optional chaining expressions in positions where short-circuiting to `undefined` causes throwing a TypeError afterward.

In your example, short-circuiting to `undefined` happens when `a` is `null` or `undefined`. In that case, only `(a?.b).c` throws an error.

```js
var a = undefined;

a?.b.c; // does not throw

(a?.b).c; // TypeError: Cannot read properties of undefined
```

The purpose of this rule is to detect possible misuse of optional chaining - cases where `?.` is used to handle nullish values but actually has no effect because the code will eventually crash when those values are nullish.

**Comment by DillonSadofsky** [comment]
I see.  You've made the policy clearer to me, and I accept that there is no linting solution that could've helped me catch the bug I wrote.  In the example I posted, the err is never nullish, but sometimes `is` will not be present.  The first case dealt only with short circuiting if the `err` symbol was nullish (since `?.` essentially targets the 'left side' of the operator, not the right side) which it never is and did nothing for cases where `is` was nullish.

I was having the (possibly malformed) opinion that if someone chains anywhere in a sequence, that is declaratively saying "at this point in the chain (and therefore everywhere to the right) a nullish value might be present" so I was thinking this linting rule would help detect a situation like the one I ran into where I got a typeerror in production code because I stopped 'early' on optional chaining.  I see now that there is no linting rule to help with this opinion.  It would be rare for me to want to be so specific to optionally chain then return to hard chaining in the same statement, so I'd personally prefer a linting rule to the effect of what I was getting at that could be manually disabled on lines where a developer specifically wishes to 'stop early' with something like my jsfiddle above:

```ts
let x = { a: undefined }

console.log(x?.a(3)) // typeerror "x.a is not a function", would not be caught by no-unsafe-chaining
console.log((x?.a)(3)) // same as line above, but WOULD be caught by no-unsafe-chaining
```
So I thought that since both of the above throw a type error and only one is caught by this rule that it would be a friendly amendment.  I'll leave this open for you to consider if such a thing is interesting to other linter users.  If not, I'll close the request.  I'll investigate the `--debug` option separately.  Its a pretty simple config that extends only `eslint:recommended` and the rule 'no-unsafe-optional-chaining' does not appear anywhere in my config, but I'll see if debug can tell me about something else affecting it.

**Comment by mdjermanovic** [comment]
It's considered a best practice to use `?.` only where it is expected that the value on its left side may be nullish.

`a?.b.c` does not imply that `a.b` may be nullish. It's perfectly valid (and recommended by the authors of the optional chaining proposal) to write `a?.b.c` where `a` is expected to be either undefined or an object with a mandatory property `b`. Without type information about `a`, a rule that requires converting `a?.b.c` to `a?.b?.c` would enforce excessive use of `?.`, which is not considered best practice.

> ```ts
> let x = { a: undefined }
> 
> console.log(x?.a(3)) // typeerror "x.a is not a function", would not be caught by no-unsafe-chaining
> console.log((x?.a)(3)) // same as line above, but WOULD be caught by no-unsafe-chaining
> ```

In this example, `?.` is used to handle nullish `x`. The rule warns on `(x?.a)(3)` because in this case `?.` doesn't handle nullish `x` well, because this code throws an error when `x` is nullish even though `?.` was used.

```js
let x = undefined

console.log(x?.a(3)) // ok, logs `undefined`
console.log((x?.a)(3)) // throws TypeError
```

**Comment by nzakas** [comment]
@mdjermanovic it seems like some of that explanation would be beneficial to add into the rule docs to help clarify its intended use?

**Comment by DillonSadofsky** [comment]
Thanks so much @mdjermanovic for taking the time to respond.  That really clarifies policy for me.  I (and the devs I work with) sometimes make the mistake of putting our optional chain handling in the wrong spot, and I thought maybe this rule would help but I understand now that that wouldn't be best practices.  Though I do think I'm probably not the only one who would maybe like such a thing (because stopping an optional chain is the exception rather than the rule), I get that its not planned.

Thanks for your time.
